File: 40-combination-sum-ii/40-combination-sum-ii.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] can, int target) {
          Arrays.sort(can);
        List<List<Integer>>res=new ArrayList<>();
        solve(0,can,target,res,new ArrayList<>(),0);
        return res;
    }
    
    public void solve(int ind,int []can,int target,List<List<Integer>>res,List<Integer>temp,int sum){
       
            if(sum==target){
                res.add(new ArrayList<>(temp));
            }
        
       
        for(int i=ind;i<can.length;i++){
            if(i!=ind && can[i]==can[i-1])continue;
            if(sum+can[i]<=target){
                sum+=can[i];
                temp.add(can[i]);
                solve(i+1,can,target,res,temp,sum);
                sum-=can[i];
                temp.remove(temp.size()-1);
            }
        }
    }
    
}

File: 40-combination-sum-ii/40-combination-sum-ii.cpp
class Solution {
public:
    
    vector<vector<int>>res;
    
    void solve(int ind,int n,vector<int> &nums,vector<int> &temp,int target){
    
            if(target==0)res.push_back(temp);
           
        
        
        for(int i=ind;i<n;i++){
            if(i!=ind && nums[i-1]==nums[i])continue;
            if(nums[i]<=target){
                temp.push_back(nums[i]);
                solve(i+1,n,nums,temp,target-nums[i]);
                temp.pop_back();
            }
        }
    }
    vector<vector<int>> combinationSum2(vector<int>& c, int target) {
        int n=c.size();
        sort(c.begin(),c.end());
        vector<int>temp;
        solve(0,n,c,temp,target);
        return res;
    }
};